webauthn-security-key/libs/auth.js

```diff
@@ -0,0 +1,694 @@
+
+// Sign out
+router.get('/signout', (req, res) => {
+  // Remove the session
+  req.session.destroy();
+  // Redirect to `/`
+  res.redirect(307, '/');
+});
+
+/**
+ * Complete the authentication
+ *
+ * Input:
+ * req.session:
+ * {
+     username: String,
+     isPasswordCorrect: String
+ * }
+ *
+ * Response as JSON:
+ * {
+     msg: String, 
+     authStatus: String // One of authStatuses 
+   }
+ * or
+ * {
+     error: String, 
+   }
+ **/
+function completeAuthentication(req, res) {
+  // username and isPasswordCorrect come from the first session, 'auth', dedicated to authentication
+  const { username, isPasswordCorrect } = req.session;
+  if (!username || !isPasswordCorrect) {
+    res.status(401).json({ error: GENERIC_AUTH_ERROR_MESSAGE });
+    return;
+  }
+  // Terminate the 'auth' session and start the 'main' session
+  // Once the 'main' session is active, the user is considered fully authenticated
+  req.session.regenerate(function (err) {
+    req.session.name = 'main';
+    // Transfer the username from the 'auth' session to the new one 'main'
+    req.session.username = username;
+  });
+  req.session.save(function (err) {
+    res.status(200).json({
+      msg: 'Authentication complete',
+      authStatus: authStatuses.COMPLETE,
+    });
+  });
+}
+
+/**
+ * Initialize the authentication: trigger completion or request a second factor, depending on the user's chosen authentication type
+ *
+ * Input:
+ * req.body:
+ * {
+     username: String,
+     password: String,
+ * }
+ * req.session:
+ * {
+     username: String,
+     ...
+ * }
+ *
+ * Response as JSON:
+ * {
+     msg: String, 
+     authStatus: String // One of authStatuses 
+   }
+ * or
+ * {
+     error: String, 
+   }
+ **/
+router.post(
+  '/initialize-authentication',
+  check('username').notEmpty().isAlphanumeric().escape(),
+  check('password').notEmpty().escape(),
+  (req, res) => {
+    // Validate the input
+    const validationErrors = validationResult(req);
+    if (!validationErrors.isEmpty()) {
+      res.status(400).json({
+        error: 'Username or password is empty or contains invalid characters',
+      });
+      return;
+    }
+    const { password, username } = req.body;
+    const user = createOrGetUser(username);
+    // Set the username in the 'auth' session so that it can be passed to the main session
+    req.session.username = username;
+    // Set the password correctness value for the next step
+    req.session.isPasswordCorrect = isPasswordCorrect(username, password);
+    // If 2FA is not set up, complete the authentication
+    const authType = getAuthType(user.credentials);
+    if (authType === authTypes.SINGLE_FACTOR) {
+      completeAuthentication(req, res);
+      // If 2FA is set up, respond with a signal that the second factor is missing
+    } else if (authType === authTypes.TWO_FACTOR) {
+      // Set the authStatus in the session so that it can be checked in server.js
```

Comment:
HTTP question:
Here I'm setting `authStatus` both:
* in `req.session` so that it can be accessible in server.js (server) 
* AND in the response so that it can be checked client-side in `index.html`
This doesn't feel optimal. Is there a better way / What am I missing?